app/middleware/cmr-granule-locator.ts

```diff
@@ -37,48 +47,73 @@ function getBbox(collection: cmr.CmrCollection, granule: cmr.CmrGranule): Boundi
+  let reason = GranuleLimitReason.None;
+  let maxResults = Number.MAX_SAFE_INTEGER;
+
+  if (req.context.serviceConfig.has_granule_limit !== false) {
+    const query = keysToLowerCase(req.query);
+    const { context } = req;
+    maxResults = env.maxGranuleLimit;
+    reason = GranuleLimitReason.System;
+
+    if ('maxresults' in query && query.maxresults < maxResults) {
+      maxResults = query.maxresults;
+      reason = GranuleLimitReason.MaxResults;
+    }
 
-  if ( req.context.serviceConfig.has_granule_limit === false ) {
-    return Number.MAX_SAFE_INTEGER;
+    const { serviceConfig } = context;
+    const serviceCollection = serviceConfig.collections?.find((sc) => sc.id === collection);
+    if (serviceCollection &&
+      serviceCollection.granuleLimit &&
+      serviceCollection.granuleLimit < maxResults) {
+      maxResults = serviceCollection.granuleLimit;
+      reason = GranuleLimitReason.Collection;
+    }
   }
 
-  let maxResults = env.maxGranuleLimit;
-  if ('maxresults' in query) {
-    maxResults = Math.min(env.maxGranuleLimit, query.maxresults);
-  }
-  return maxResults;
+  return [maxResults, reason];
 }
 
 /**
- * Returns a warning message if not all matching granules will be processed for the request
- *
+ * Create a message indicating that the results have been limited and why - if necessary
+ * 
+ * @param req - The client request, containing an operation
+ * @param collection - The id of the collection to which the granules belong
  * @returns a warning message if not all matching granules will be processed, or undefined
- * if not applicable
+ * if not applicable 
  */
-function getResultsLimitedMessage(req: HarmonyRequest): string {
+function getResultsLimitedMessage(req: HarmonyRequest, collection: string): string {
   const { operation } = req;
   let message;
 
   if ( req.context.serviceConfig.has_granule_limit == false ) return message;
 
-  let numGranules = operation.cmrHits;
-  if (operation.maxResults) {
-    numGranules = Math.min(numGranules, operation.maxResults, env.maxGranuleLimit);
-  } else {
-    numGranules = Math.min(numGranules, env.maxGranuleLimit);
-  }
+  const [numGranules, reason] = getMaxGranules(req, collection);
 
   if (operation.cmrHits > numGranules) {
     message = `CMR query identified ${operation.cmrHits} granules, but the request has been limited `
      + `to process only the first ${numGranules} granules`;
-    if (operation.maxResults && operation.maxResults < env.maxGranuleLimit) {
-      message += ` because you requested ${operation.maxResults} maxResults.`;
-    } else {
-      message += ' because of system constraints.';
+
+    switch (reason) {
+      case GranuleLimitReason.MaxResults:
+        message += ` because you requested ${operation.maxResults} maxResults.`;
+        break;
+
+      case GranuleLimitReason.Collection:
+        message += ` because collection ${collection} is limited to ${numGranules} for the ${req.context.serviceConfig.name} service.`;
```

Comment:
I'm not sure I'd put the service name in there, as it feels like sausage making. Maybe just "for your requested operations" or something like that.

Reply:
I think based on the reasons Chris mentioned above we need to add the service name - the same collection might have different limits defined for different services so we can just say "limited for the collection" as it would be confusing if two different limits are mentioned for the same collection without distinguishing by service as well.